Backport gnulib's COUNTED_BY annotation for the newer gettext sources.
Keep the upstream compiler guard: GCC does not yet support pointer fields.

--- a/gnulib/lib/attribute.h
+++ b/gnulib/lib/attribute.h
@@ -208,6 +208,19 @@
    following 'case' or 'default' label.  */
 /* Applies to: Empty statement (;), inside a 'switch' statement.  */
 #define FALLTHROUGH _GL_ATTRIBUTE_FALLTHROUGH
+
+
+/* =================== Attributes for runtime diagnostics =================== */
+
+/* Attributes that provide information to the undefined-behaviour sanitizer
+   (UBSAN).  */
+
+/* COUNTED_BY (C) declares that the number of elements of the field is given
+   by C, which must be another field in the same struct.
+   The programmer is responsible for guaranteeing some invariants; see
+   <https://gcc.gnu.org/onlinedocs/gcc/Common-Attributes.html> for details.  */
+/* Applies to struct fields of type array or pointer (to data).  */
+#define COUNTED_BY(c) _GL_ATTRIBUTE_COUNTED_BY (c)


 /* ================== Attributes for debugging information ================== */
--- a/gnulib/m4/gnulib-common.m4
+++ b/gnulib/m4/gnulib-common.m4
@@ -433,7 +433,24 @@
 #  define _GL_ATTRIBUTE_CONST __attribute__ ((__const__))
 # else
 #  define _GL_ATTRIBUTE_CONST
+# endif
+#endif
+
+/* _GL_ATTRIBUTE_COUNTED_BY (C) declares that the number of elements of
+   the field is given by C, which must be another field in the same struct.
+   The programmer is responsible for guaranteeing some invariants; see
+   <https://gcc.gnu.org/onlinedocs/gcc/Common-Attributes.html> for details.  */
+/* Applies to struct fields of type array or pointer (to data).  */
+#ifndef _GL_ATTRIBUTE_COUNTED_BY
+/* This attributes is supported
+     - for fields of array type: by gcc >= 16, clang >= 18,
+     - for fields of pointer type: by gcc when <https://gcc.gnu.org/PR125072>
+       will be fixed, clang >= 19.  */
+# if defined __clang__ && __clang_major__ >= 19
+#  define _GL_ATTRIBUTE_COUNTED_BY(c) __attribute__ ((__counted_by__ (c)))
+# else
+#  define _GL_ATTRIBUTE_COUNTED_BY(c)
 # endif
 #endif

 /* _GL_ATTRIBUTE_DEALLOC (F, I) declares that the function returns pointers
